### PR TITLE
update: stash save/pop uncommitted changes.

### PR DESCRIPTION
**Pull request by mikemcquaid**
Also:
- return to your previous branch after `brew update`.

CC @Homebrew/owners for thoughts/review.

### CODE REVIEW

**Comment by mikemcquaid** [comment]
I think this will help cut down the number of problems we see when these files have randomly changed.

**Comment by DomT4** [comment]
What happens if I'm on a branch rather than ` master `? Does it go ahead and stash any changes I've made? Can we scope it to ` master ` or guard it against things like that by only running it if ` HOMEBREW_DEVELOPER ` isn't set? Just worried about having to ` pop ` the branch pretty regularly when I forget to switch back to ` master ` because I'm an idiot.

**Comment by mikemcquaid** [comment]
We could scope this just to `master` but the alternative is that `brew update` fails. Personally I just use `git fetch` or `git pull` to update normally.

**Comment by bfontaine** [comment]
Why don’t you automatically pop changes after the update? It could cause merge conflicts on pop, but that’s what [`git-up`](https://github.com/aanand/git-up) is doing, and if one misses the message they’ll wonder where their changes are.

**Comment by dunn** [comment]
> We could scope this just to `master` but the alternative is that `brew update` fails.

Personally I'd be happy with `brew update` failing when I'm not on the `master` branch.

**Comment by mikemcquaid** [comment]
> Why don’t you automatically pop changes after the update? It could cause merge conflicts on pop, but that’s what git-up is doing, and if one misses the message they’ll wonder where their changes are.

Yeh, I'm less concerned about that. If people are making changes and not committing them they're going to have other problems. I dunno, though.

**Comment by mikemcquaid** [comment]
Any other @Homebrew/owners thoughts here?

**Comment by asparagui** [comment]
I use brew update to get back to master, but that's probably not a good usecase.

**Comment by mikemcquaid** [comment]
I've updated this to also `stash pop` at the end, tell the user that and what we are stashing and defaulting to `git pull --rebase` (although happy to move that to another PR). @Homebrew/owners thoughts?

**Comment by xu-cheng** [comment]
Have you tested how it will behave when dealing with interrupt?

**Comment by mikemcquaid** [comment]
Good point. How do you think it should behave on an interrupt? Restore the stash?

**Comment by xu-cheng** [comment]
I dunno. It's kinda pain to handle interrupt as I have seen when I implemented multithread brew update.

I guess if the interrupt happens previous `git pull`, it should restore. After `git pull` finishes, interrupt should not be allowed.

File: Library/Homebrew/cmd/update.rb [hunk 1]
```diff
@@ -122,25 +122,41 @@ class Updater
...
+    args = ["pull", "--rebase", "origin"]
     # the refspec ensures that 'origin/master' gets updated
     args << "refs/heads/master:refs/remotes/origin/master"
+    args += quiet
 
     reset_on_interrupt { safe_system "git", *args }
 
+    if @stashed
+      puts "Restored your changes:"
+      safe_system "git", "stash", "pop", *quiet
```
Comment (xu-cheng) [inline Library/Homebrew/cmd/update.rb]: I think you should checkout the original branch as well if it isn't master.

Comment (mikemcquaid) [inline Library/Homebrew/cmd/update.rb]: Yeh, it's tricky as @asparagui has said he explicitly relies on the current behaviour to change back to `master`.

Comment (xu-cheng) [inline Library/Homebrew/cmd/update.rb]: In that case, we should checkout original branch, `git stash pop` then checkout master.

Comment (mikemcquaid) [inline Library/Homebrew/cmd/update.rb]: That will make the stash more likely to succeed but the checkout more likely to fail.

**Comment by mikemcquaid** [comment]
@Homebrew/owners would anyone object (or prefer!) if this meant that `brew update` switched back to your work branch after doing a `brew update`?

**Comment by mikemcquaid** [comment]
I've pushed the latest version that checks out the old branch again, handles interrupts and removes the `rebase` stuff. Let me know any more thoughts @Homebrew/owners.

**Comment by mikemcquaid** [comment]
Any final thoughts on this @Homebrew/owners? If not I'll probably merge fairly shortly.

**Comment by xu-cheng** [comment]
LGTM :+1:

**Comment by atdt** [comment]
@mikemcquaid, just wanted to express my appreciation for this work -- it makes working with Homebrew even more frictionless and pleasurable.

**Comment by mikemcquaid** [comment]
Thanks @atdt for taking the time to say nice things :sparkles:

**Pull request closed without merging**